### PR TITLE
Send Icinga2 endpoint config updates

### LINKED ISSUE #7: Send config::Update
* Add new CLI parameters specifying a zone name and an amount of files
* To every client send a [config::Update message](https://github.com/Icinga/icinga2/blob/c788709593d7d740295b0eb89e7cb171dfab6ce7/lib/remote/apilistener-filesync.cpp#L221-L229) (once) with config for the specified zone containing a random amount [0..CLIparam] of empty files (1.conf, 2.conf, ...)

### PR DESCRIPTION
**Pull request by yhabteab**
resolves #7

### CODE REVIEW

**Comment by yhabteab** [comment]
Testprotocol:

Icinga2 Agent logs:

```bash
2020/12/10 10:04:21 SERVER: Started listening on '0.0.0.0:443'
2020/12/10 10:04:32 SERVER: Accepted connection from 127.0.0.1:57573
2020/12/10 10:04:32 SERVER: Connection approved from Server
2020/12/10 10:04:32 SERVER: Connection: Waiting for the client
2020/12/10 10:04:32 SERVER: JsonRpcConnection: Received "icinga::Hello" from the client
2020/12/10 10:04:32 SERVER: JsonRpcConnection: Sending config::Update JSRPC message: {"jsonrpc":"2.0","method":"config::Update","params":{"update":{"ha-cluster":{"/_etc/":"downtimes.conf"}},"update_v2":{"ha-cluster":{"/_etc/":"hosts.conf"}},"checksums":{"ha-cluster":{"/_etc/":"commands.conf"}}}}
2020/12/10 10:04:32 SERVER: JsonRpcConnection: Waiting for rand generating times to send 4187055439422133645
2020/12/10 10:05:42 SERVER: Accepted connection from 127.0.0.1:57586
2020/12/10 10:05:42 SERVER: Connection approved from Server
2020/12/10 10:05:42 SERVER: Connection: Waiting for the client
2020/12/10 10:05:42 SERVER: JsonRpcConnection: Received "icinga::Hello" from the client
2020/12/10 10:05:42 SERVER: JsonRpcConnection: Sending config::Update JSRPC message: {"jsonrpc":"2.0","method":"config::Update","params":{"update":{"ha-cluster":{"/_etc/":"downtimes.conf"}},"update_v2":{"ha-cluster":{"/_etc/":"hosts.conf"}},"checksums":{"ha-cluster":{"/_etc/":"commands.conf"}}}}
2020/12/10 10:05:42 SERVER: JsonRpcConnection: Waiting for rand generating times to send 5954857764366209179
```

Test server logs:

```bash
2020/12/10 13:41:33 SERVER: Started listening on '0.0.0.0:443'
2020/12/10 13:41:37 SERVER: Accepted connection from 127.0.0.1:60402
2020/12/10 13:41:37 SERVER: Connection approved from Server
2020/12/10 13:41:37 SERVER: Connection: Waiting for the client
2020/12/10 13:41:37 SERVER: JsonRpcConnection: Received "icinga::Hello" from the client
2020/12/10 13:41:37 SERVER: JsonRpcConnection: Sending config::Update JSRPC message:
{
   "jsonrpc":"2.0",
   "method":"config::Update",
   "params":{
      "update":{
         "ha-cluster":{
            "156.conf":"156.conf",
            "1737.conf":"1737.conf",
            "2451.conf":"2451.conf",
            "2888.conf":"2888.conf",
            "3000.conf":"3000.conf",
            "3237.conf":"3237.conf",
            "3274.conf":"3274.conf",
            "3300.conf":"3300.conf",
            "3721.conf":"3721.conf",
            "4059.conf":"4059.conf",
            "4147.conf":"4147.conf",
            "4425.conf":"4425.conf",
            "4538.conf":"4538.conf",
            "5026.conf":"5026.conf",
            "5094.conf":"5094.conf",
            "5194.conf":"5194.conf",
            "5466.conf":"5466.conf",
            "5541.conf":"5541.conf",
            "5561.conf":"5561.conf",
            "5746.conf":"5746.conf",
            "6159.conf":"6159.conf",
            "6831.conf":"6831.conf",
            "7996.conf":"7996.conf",
            "8047.conf":"8047.conf",
            "8081.conf":"8081.conf",
            "8162.conf":"8162.conf",
            "8643.conf":"8643.conf",
            "9002.conf":"9002.conf",
            "9241.conf":"9241.conf",
            "953.conf":"953.conf"
         }
      },
      "update_v2":{
         "ha-cluster":{
            "1137.conf":"1137.conf",
            "1211.conf":"1211.conf",
            "1353.conf":"1353.conf",
            "1528.conf":"1528.conf",
            "1563.conf":"1563.conf",
            "1577.conf":"1577.conf",
            "2081.conf":"2081.conf",
            "2540.conf":"2540.conf",
            "2790.conf":"2790.conf",
            "3891.conf":"3891.conf",
            "4078.conf":"4078.conf",
            "408.conf":"408.conf",
            "5089.conf":"5089.conf",
            "5429.conf":"5429.conf",
            "563.conf":"563.conf",
            "59.conf":"59.conf",
            "631.conf":"631.conf",
            "6413.conf":"6413.conf",
            "6420.conf":"6420.conf",
            "694.conf":"694.conf",
            "7189.conf":"7189.conf",
            "7202.conf":"7202.conf",
            "7887.conf":"7887.conf",
            "8266.conf":"8266.conf",
            "8510.conf":"8510.conf",
            "8705.conf":"8705.conf",
            "9106.conf":"9106.conf",
            "9703.conf":"9703.conf",
            "9718.conf":"9718.conf",
            "9947.conf":"9947.conf"
         }
      },
      "checksums":{
         "ha-cluster":{
            "1318.conf":"1318.conf",
            "1445.conf":"1445.conf",
            "1485.conf":"1485.conf",
            "1847.conf":"1847.conf",
            "1957.conf":"1957.conf",
            "2002.conf":"2002.conf",
            "2199.conf":"2199.conf",
            "2433.conf":"2433.conf",
            "2605.conf":"2605.conf",
            "2888.conf":"2888.conf",
            "3015.conf":"3015.conf",
            "3033.conf":"3033.conf",
            "3090.conf":"3090.conf",
            "3133.conf":"3133.conf",
            "4324.conf":"4324.conf",
            "4376.conf":"4376.conf",
            "456.conf":"456.conf",
            "4728.conf":"4728.conf",
            "4783.conf":"4783.conf",
            "495.conf":"495.conf",
            "5356.conf":"5356.conf",
            "5447.conf":"5447.conf",
            "6258.conf":"6258.conf",
            "7387.conf":"7387.conf",
            "7463.conf":"7463.conf",
            "8287.conf":"8287.conf",
            "8511.conf":"8511.conf",
            "8623.conf":"8623.conf",
            "9355.conf":"9355.conf",
            "9828.conf":"9828.conf"
         }
      }
   }
}
```

File: main.go [hunk 1]
```diff
@@ -115,6 +131,56 @@ func ProcessIncomingConnection(conn net.Conn) {
 
 		if decoder.Method == "icinga::Hello" {
 			log.Printf("SERVER: JsonRpcConnection: Received %q from the client\n", decoder.Method)
+
+			math.Seed(time.Now().UTC().UnixNano())
+
+			for i := 0; i < *files; i++ {
+				encoder := &Encoder{
+					JSONRPC: "2.0",
+					Method: "config::Update",
```
Comment (Al2Klimov) [inline main.go]: The task was to send **one** config::Update...

File: main.go [hunk 2]
```diff
@@ -115,6 +131,56 @@ func ProcessIncomingConnection(conn net.Conn) {
...
+							},
+						},
+						Checksums: map[string]ConfigInfo{
+							*zone: {
+								"/_etc/": "commands.conf",
+								"/.checksums":"fb2b4a66074cfc8a6aadc640fee441b46f85f116140fa39c405c6ef67b199c0f",
+								"/.timestamp":"1c3b632eebb230cb77f7bfaf74f3380b62f99bb672d05cca1298917260d3f85f",
+								"/_etc/app.conf":"006036009ed0e202a3bae7eaf9018ae21adf8a7a628caa79797740fc18805460",
+							},
+						},
```
Comment (Al2Klimov) [inline main.go]: ... with a **random amount** of files.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: main.go (hunk 2 above)
Comment (Al2Klimov) [reply main.go]: And this one?

File: main.go [hunk 3]
```diff
@@ -115,17 +131,84 @@ func ProcessIncomingConnection(conn net.Conn) {
...
+
+				encoder.Params.Update_v2[*zone][path] = path
+
+				random, err := GenerateFile("conf")
+
+				if err != nil {
+					log.Fatalf("SERVER: JsonRpcConnection: Error while generating new config files: %s", err)
+				}
+
+				encoder.Params.Checksums[*zone][random] = random
```
Comment (Al2Klimov) [inline main.go]: * According to #7 the files shall be empty
* The file names shall be the same for Update, Update_v2 and Checksums

File: main.go [hunk 4]
```diff
@@ -115,17 +131,84 @@ func ProcessIncomingConnection(conn net.Conn) {
...
+	if *val == " " {
+		log.Fatalf("SERVER: Required command line '%s' argment is empty\n", desc)
+	}
+}
+
+func GenerateFile(suffix string) (string, error) {
+	r := math.Intn(1000000)
+	for i := 0; i < 10; i++ {
+		fname := fmt.Sprintf("%d.%s", r, suffix)
+		if _, err := os.Stat(fname); os.IsNotExist(err) {
```
Comment (Al2Klimov) [inline main.go]: What's this condition for?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: main.go (hunk 4 above)
Comment (yhabteab) [reply main.go]: To verify that there are no errors when creating the file.

Comment (Al2Klimov) [reply main.go]: That's unnecessary.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
How the Icinga 2 agent behaves (logs) with this one being the master (test protocol)?

**Comment by yhabteab** [comment]
**Test server and Icinga Agent**:

Test server logs:

```bash
2020/12/10 14:26:13 SERVER: Started listening on '0.0.0.0:443'
2020/12/10 14:26:20 SERVER: Accepted connection from 127.0.0.1:60810
2020/12/10 14:26:20 SERVER: Connection approved from Server
2020/12/10 14:26:20 SERVER: Connection: Waiting for the client
2020/12/10 14:26:20 SERVER: JsonRpcConnection: Received "icinga::Hello" from the client
2020/12/10 14:26:21 SERVER: JsonRpcConnection: Sending config::Update JSRPC message: 
{"jsonrpc":"2.0","method":"config::Update","params":{"update":{"ha-cluster":{"10002.conf":" ","100384.conf":" 
[...]
            }
        }
    }
}
```

Agent logs:

```bash
[...]
[2020-12-10 14:26:34 +0100] information/ApiListener: Stage: Updating received configuration file '/Users/yhabteab/Workspace/cluster/prefix/var/lib/icinga2/api/zones-stage/ha-cluster/999869.conf' for zone 'ha-cluster'.
[2020-12-10 14:26:34 +0100] information/ApiListener: Stage: Updating received configuration file '/Users/yhabteab/Workspace/cluster/prefix/var/lib/icinga2/api/zones-stage/ha-cluster/999982.conf' for zone 'ha-cluster'.
[2020-12-10 14:26:34 +0100] information/ApiListener: Applying configuration file update for path '/Users/yhabteab/Workspace/cluster/prefix/var/lib/icinga2/api/zones-stage/ha-cluster' (8045 Bytes).
[2020-12-10 14:26:34 +0100] information/ApiListener: Received configuration updates (1) from endpoint 'ha-master' are different to production, triggering validation and reload.
[2020-12-10 14:26:34 +0100] notice/Process: Running command '/Users/yhabteab/Workspace/cluster/prefix/lib/icinga2/sbin/icinga2' 'daemon' '-x' 'notice' '--validate' '--define' 'System.ZonesStageVarDir=/Users/yhabteab/Workspace/cluster/prefix/var/lib/icinga2/api/zones-stage/': PID 32148
[2020-12-10 14:26:35 +0100] notice/CheckerComponent: Pending checkables: 0; Idle checkables: 0; Checks/s: 0
[2020-12-10 14:26:40 +0100] notice/CheckerComponent: Pending checkables: 0; Idle checkables: 0; Checks/s: 0
[2020-12-10 14:26:40 +0100] notice/ApiListener: Updating object authority for objects at endpoint 'ha-cluster'.
[2020-12-10 14:26:40 +0100] notice/ApiListener: Current zone master: ha-cluster
[2020-12-10 14:26:40 +0100] notice/ApiListener: Connected endpoints: ha-master (1)
[...]
```

**Tests with Icinga2 Master & Agent**:

Icinga2 Master logs:

```bash
[2020-12-10 14:36:15 +0100] information/ApiListener: Sending config updates for endpoint 'ha-cluster' in zone 'ha-cluster'.
[2020-12-10 14:36:15 +0100] information/ApiListener: Syncing configuration files for zone 'ha-cluster' to endpoint 'ha-cluster'.
[2020-12-10 14:36:15 +0100] information/ApiListener: Syncing Json encoded configuration to the endpoint {"jsonrpc":"2.0","method":"config::Update","params":{"checksums":{"ha-cluster":{"/.checksums":"fb2b4a66074cfc8a6aadc640fee441b46f85f116140fa39c405c6ef67b199c0f",

[...]
         }
      }
   }
}

[2020-12-10 14:36:15 +0100] information/ApiListener: Finished sending config file updates for endpoint 'ha-cluster' in zone 'ha-cluster'.
[2020-12-10 14:36:15 +0100] information/ApiListener: Syncing runtime objects to endpoint 'ha-cluster'.
[2020-12-10 14:36:15 +0100] information/ApiListener: Finished syncing runtime objects to endpoint 'ha-cluster'.
[2020-12-10 14:36:15 +0100] information/ApiListener: Finished sending runtime config updates for endpoint 'ha-cluster' in zone 'ha-cluster'.
[2020-12-10 14:36:15 +0100] information/ApiListener: Sending replay log for endpoint 'ha-cluster' in zone 'ha-cluster'.
[2020-12-10 14:36:15 +0100] information/ApiListener: Finished sending replay log for endpoint 'ha-cluster' in zone 'ha-cluster'.
[2020-12-10 14:36:15 +0100] information/ApiListener: Finished syncing endpoint 'ha-cluster' in zone 'ha-cluster'
```

Agent logs:

```bash
[2020-12-10 14:36:15 +0100] notice/JsonRpcConnection: Received 'icinga::Hello' message from identity 'ha-master'.
[2020-12-10 14:36:15 +0100] notice/JsonRpcConnection: Received 'config::Update' message from identity 'ha-master'.
[2020-12-10 14:36:15 +0100] information/ApiListener: Applying config update from endpoint 'ha-master' of zone 'ha-master'.
[2020-12-10 14:36:15 +0100] information/ApiListener: Received configuration for zone 'ha-cluster' from endpoint 'ha-master'. Comparing the timestamp and checksums.
[2020-12-10 14:36:15 +0100] information/ApiListener: Stage: Updating received configuration file '/Users/yhabteab/Workspace/cluster/prefix/var/lib/icinga2/api/zones-stage/ha-cluster//_etc/app.conf' for zone 'ha-cluster'.
[2020-12-10 14:36:15 +0100] information/ApiListener: Stage: Updating received configuration file '/Users/yhabteab/Workspace/cluster/prefix/var/lib/icinga2/api/zones-stage/ha-cluster//_etc/commands.conf' for zone 'ha-cluster'.

[...]

[2020-12-10 14:36:15 +0100] information/ApiListener: Stage: Updating received configuration file '/Users/yhabteab/Workspace/cluster/prefix/var/lib/icinga2/api/zones-stage/ha-cluster//_etc/users.conf' for zone 'ha-cluster'.
[2020-12-10 14:36:15 +0100] information/ApiListener: Applying configuration file update for path '/Users/yhabteab/Workspace/cluster/prefix/var/lib/icinga2/api/zones-stage/ha-cluster' (15628 Bytes).
[2020-12-10 14:36:15 +0100] information/ApiListener: Received configuration updates (1) from endpoint 'ha-master' are different to production, triggering validation and reload.
[2020-12-10 14:36:15 +0100] notice/Process: Running command '/Users/yhabteab/Workspace/cluster/prefix/lib/icinga2/sbin/icinga2' 'daemon' '-x' 'notice' '--validate' '--define' 'System.ZonesStageVarDir=/Users/yhabteab/Workspace/cluster/prefix/var/lib/icinga2/api/zones-stage/': PID 32775
```

**Comment by Al2Klimov** [comment]
What's an **actual Icinga 2 master** doing there?

**Comment by Al2Klimov** [comment]
Is the agent reloading only once?

**Comment by yhabteab** [comment]
Yes, once after the received config are finished being validated.

**Comment by Al2Klimov** [comment]
And then it re-connects to the test server and doesn't reload?

**Comment by yhabteab** [comment]
If it does not get a configuration change no it will not reloaded!

```bash
[2020-12-10 15:45:20 +0100] information/ApiListener: Applying configuration file update for path '/Users/yhabteab/Workspace/cluster/prefix/var/lib/icinga2/api/zones-stage/ha-cluster' (8045 Bytes).
[2020-12-10 15:45:20 +0100] information/ApiListener: Received configuration updates (1) from endpoint 'ha-master' are different to production, triggering validation and reload.
[2020-12-10 15:45:20 +0100] notice/Process: Running command '/Users/yhabteab/Workspace/cluster/prefix/lib/icinga2/sbin/icinga2' 'daemon' '-x' 'notice' '--validate' '--define' 'System.ZonesStageVarDir=/Users/yhabteab/Workspace/cluster/prefix/var/lib/icinga2/api/zones-stage/': PID 35440
[2020-12-10 15:45:24 +0100] notice/ApiListener: Updating object authority for objects at endpoint 'ha-cluster'.
```

```bash
[2020-12-10 15:46:45 +0100] information/ApiListener: Reconnecting to endpoint 'ha-master' via host '127.0.0.1' and port '443'
[2020-12-10 15:46:45 +0100] notice/ApiListener: Connected endpoints: 
[2020-12-10 15:46:45 +0100] information/ApiListener: New client connection for identity 'ha-master' to [127.0.0.1]:443
[2020-12-10 15:46:45 +0100] notice/ApiListener: New JSON-RPC client
[2020-12-10 15:46:45 +0100] information/ApiListener: Requesting new certificate for this Icinga instance from endpoint 'ha-master'.
[2020-12-10 15:46:45 +0100] information/ApiListener: Sending config updates for endpoint 'ha-master' in zone 'ha-master'.
[2020-12-10 15:46:45 +0100] information/ApiListener: Finished sending config file updates for endpoint 'ha-master' in zone 'ha-master'.
[2020-12-10 15:46:45 +0100] information/ApiListener: Syncing runtime objects to endpoint 'ha-master'.
[2020-12-10 15:46:45 +0100] information/ApiListener: Finished syncing runtime objects to endpoint 'ha-master'.
[2020-12-10 15:46:45 +0100] information/ApiListener: Finished sending runtime config updates for endpoint 'ha-master' in zone 'ha-master'.
[2020-12-10 15:46:45 +0100] information/ApiListener: Sending replay log for endpoint 'ha-master' in zone 'ha-master'.

[2020-12-10 15:46:45 +0100] information/ApiListener: Finished sending replay log for endpoint 'ha-master' in zone 'ha-master'.
[2020-12-10 15:46:45 +0100] information/ApiListener: Finished syncing endpoint 'ha-master' in zone 'ha-master'.
[2020-12-10 15:46:45 +0100] information/ApiListener: Finished reconnecting to endpoint 'ha-master' via host '127.0.0.1' and port '443'
```

**Comment by Al2Klimov** [comment]
But the test server's Icinga configuration changes randomly. That shall cause the agent to reload over and over again.

**Review by Al2Klimov — APPROVED** [review]

**Pull request merged**